packages/go/bomenc/encodings.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2024 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package bomenc
+
+// Encoding interface defines the methods that all encoding types must implement.
+// This interface provides a unified way to handle different encodings throughout the package,
+// allowing us to treat all encodings uniformly. This design facilitates easy extension
+// and manipulation of different encoding types without altering the core logic.
+type Encoding interface {
+	// Sequence returns the byte sequence that represents the Byte Order Mark (BOM) for this encoding.
+	// This method is crucial for identifying the specific byte sequence that indicates
+	// this encoding at the start of a file.
+	Sequence() []byte
+
+	// String returns a human-readable string representation of the encoding.
+	// This is particularly useful for logging and user interfaces, providing
+	// a user-friendly name for the encoding.
+	String() string
+
+	// HasSequence checks if the given data starts with this encoding's BOM sequence.
+	// This method allows for efficient checking of whether a given byte slice
+	// begins with this encoding's BOM, which is essential for encoding detection.
+	HasSequence(data []byte) bool
+}
+
+// bomEncoding is the concrete implementation of the Encoding interface.
+// It encapsulates all necessary information and behavior for a specific encoding,
+// providing a consistent structure for handling different encodings. This approach
+// allows us to create instances for each supported encoding while maintaining
+// a uniform interface for interaction.
+type bomEncoding struct {
+	encodingType    string                 // A human-readable name for the encoding
+	sequence        []byte                 // The BOM sequence for this encoding
+	hasSequenceFunc func(data []byte) bool // Function to check if data starts with this encoding's BOM
+}
+
+// String returns the human-readable name of the encoding.
+// This method fulfills the Encoding interface and provides a simple way
+// to get a string representation of the encoding.
+func (b bomEncoding) String() string {
```

Comment:
As per our golang standards all receivers should be `s`. I see this in a few places but will only write this comment for it.